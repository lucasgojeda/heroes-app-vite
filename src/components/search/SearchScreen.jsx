import React, { useMemo } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';




export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const {q = ''} = queryString.parse(location.search);

    

    const [ { searchText }, handleInputChange, reset ] = useForm( {
        searchText: q
    } );

    const heroesFiltered =  useMemo( () => (getHeroesByName(q)), [q]);

    const handleSearch = (e) => {
        e.preventDefault();

        navigate(`?q=${ searchText }`)

        
    }

    return(
        <>
            <h1> Búsquedas </h1>
            <hr />

            <div className='row'>
                <div className='col-5'>
                    <h4>Buscar</h4>
                    <hr />

                    <form
                        onSubmit={ handleSearch }
                    >
                        <input 
                            onChange={handleInputChange}
                            type='text'
                            placeholder='Buscar un heroe'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'
                            value={ searchText }
                        />

                        <button
                            type='submit'
                            className='btn btn-outline-primary mt-1'
                        >
                            Buscar...
                        </button>
                    </form>
                </div>
                <div className='col-7'>
                    <h4>Resultados</h4>
                    <hr />

                    {
                        (q === '')
                            ? <div className='alert alert-info'> Buscar un héroe </div>
                            : ( heroesFiltered.length === 0 ) && <div className='alert alert-danger'> No hay Resultados de '{q}' </div>
                    }

                    {
                        heroesFiltered.map(hero => (
                            <HeroCard 
                                key={hero.id}
                                { ...hero }
                            />
                        ))
                    }
                </div>
            </div>

            
        </>
    );
};