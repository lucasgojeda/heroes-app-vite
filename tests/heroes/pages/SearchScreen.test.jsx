import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { SearchScreen } from '../../../src/components/search/SearchScreen';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate,
}));



describe('Pruebas en <SearchScreen />', () => {


    beforeEach(() => jest.clearAllMocks() );

    
    test('debe de mostrarse correactamente con valores por defecto', () => {
        
        const { container } =render(
            <MemoryRouter>
                <SearchScreen />
            </MemoryRouter>
        );
        expect( container ).toMatchSnapshot();
        
    });

    test('debe de mostrar a Batman y el input con el valor del queryString', () => {
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        ); 
        
        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('batman');
        
        const img = screen.getByRole('img');
        expect( img.src ).toContain('https://res.cloudinary.com/the-kings-company/image/upload/v1658331093/heroes-app/dc-batman_qujtsk.jpg');

        const alert = screen.getByLabelText('alert-danger');
        expect( alert.style.display ).toBe('none');
    });

    test('debe de mostrar un error si no se encuentra el hero (batman123)', () => {
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchScreen />
            </MemoryRouter>
        );

        const alert = screen.getByLabelText('alert-danger');
        expect( alert.style.display ).toBe('');
        

    });

    test('debe de llamar el navigate a la pantalla nueva', () => {
        
        const inputValue = 'superman';

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change( input, { target: { name: 'searchText', value: inputValue }})
        
        
        const form = screen.getByRole('form');
        fireEvent.submit( form );
        
        expect( mockedUseNavigate ).toHaveBeenCalledWith(`?q=${ inputValue }`)

    });


});