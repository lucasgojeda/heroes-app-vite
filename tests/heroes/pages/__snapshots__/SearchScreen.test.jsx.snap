// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrarse correactamente con valores por defecto 1`] = `
<div>
  <h1>
     Búsquedas 
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        name="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Buscar un heroe"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        class="alert alert-info"
      >
        Buscar un héroe
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger"
        style="display: none;"
      >
        No hay Resultados de '
        '
      </div>
    </div>
  </div>
</div>
`;
